ruby/ext/google/protobuf_c/message.c

```diff
@@ -734,7 +734,16 @@ uint64_t Message_Hash(const upb_msg* msg, const upb_msgdef* m, uint64_t seed) {
  */
 static VALUE Message_hash(VALUE _self) {
   Message* self = ruby_to_Message(_self);
-  return INT2FIX(Message_Hash(self->msg, self->msgdef, 0));
+  uint64_t hash_value = Message_Hash(self->msg, self->msgdef, 0);
+  // Fixnum is restricted to signed long, so downcast if necessary.
+  long long_value = (long)hash_value;
```

Comment:
What about:

```
  // RUBY_FIXNUM_MAX should be one less than a power of 2.
  assert(RUBY_FIXNUM_MAX & (RUBY_FIXNUM_MAX + 1) == 0);  
  return INT2FIX(hash_value & RUBY_FIXNUM_MAX);
```

Reply:
Yeah, that works. It restricts hash values to positive, but I think 32-bit hashes are always positive for most Ruby objects anyway.